Policy enhancements

tgraf opened the pull request:
Added by aanm:
Depends on #650
Fixes https://github.com/cilium/cilium/issues/552
TODO:
- [x] Fix runtime tests. They are failing most likely due `policy: changed TargetCoveredBy to cover all labels` (134fafe).
- [x] Merge ingress-isolation code with this PR
- [x] Add namespace listener to automatically modify the running containers' labels in case the user changes the label of a kubernetes namespace.
- [ ] Rebase with #650
- [x] Add documentation to make @houndci-bot happy

aanm commented on pkg/policy/node.go:
```diff
@@ -26,11 +26,16 @@ import (
+	// CoverAll marks policy nodes which should always have their
+	// rules applied regardless of whether the name/path of the node
+	// matches any of the labels in the search context.
+	CoverAll  bool `json:"coverAll,omitempty"`
```
It took me a while to figure out the meaning of this.
IMO `CoverAll` means the node should cover **all** coverages in order to apply the rules.
This is the requirement for the kubernetes network policy labels [here](https://github.com/kubernetes/apimachinery/blob/9c89140a45325f83409e90359888133478207c7a/pkg/apis/meta/v1/types.go#L767)
For the first pod selector we will have 1 or more coverages in a single node, for example:
```
"spec": {
    "podSelector": {
      "matchLabels": {
        "guestbook": "web",
        "k8s-app": "kubedns"
      }
    },
    "ingress": [
      {
        "from": [
          {
            "podSelector": {
              "matchLabels": {
                "io.cilium.reserved": "world"
              }
            }
          }
        ]
      }
    ]
  }
```
Will result into:
```
"guestbook-web": {
    "name": "guestbook-web",
    "rules": [
        {
            "coverage": [
                {
                    "key": "guestbook",
                    "value": "web",
                    "source": "k8s"
                },
                {
                    "key": "k8s-app",
                    "value": "kubedns",
                    "source": "k8s"
                }
            ],
            "allow": [
                {
                    "action": "always-accept",
                    "matchLabels": [
                        {
                            "key": "world",
                            "source": "reserved"
                        }
                    ]
                }
            ]
        }
    ],
    "coverAll": true
}
```

aanm replied:
With coverAll=true this test fails but it should pass.
```
	ctx = policy.SearchContext{
		From: labels.LabelArray{
			labels.NewLabel("foo4", "bar4", common.K8sLabelSource),
			labels.NewLabel("foo3", "bar3", common.K8sLabelSource),
		},
		To: labels.LabelArray{
			labels.NewLabel("foo2", "bar2", common.K8sLabelSource),
		},
		Trace: policy.TRACE_VERBOSE,
	}
	c.Assert(node.Allows(&ctx), Equals, api.DENY)
```
I will change the `coverAll` behavior to my point of view. Makes sense?

tgraf replied:
The mean of `coverAll` is that the name of the policy node should not be a requirement for any of the rules of the node to be considered. The coverage of the rules is defined by the coverage of the rules themselves. The node itself is a pure "packaging" construct. This allows to map k8s NetworkPolicy to policy nodes without requiring the name of the policy node to somehow correlate to the coverage of the actual rules.

I think you are saying the same thing but wanted to make sure.

tgraf commented on pkg/policy/node.go:
```diff
@@ -26,16 +26,16 @@ import (
+	// CoverAll marks policy nodes which should always have their
+	// rules applied regardless of whether the name/path of the node
+	// matches any of the labels in the search context.
+	CoverAll bool             `json:"coverAll,omitempty"`
```
I assume you will squash this change with the intro patch of this field.

aanm replied:
Done, I squashed the whole commit since it was that change + documentation

tgraf commented on pkg/policy/policy.go:
```diff
@@ -140,19 +140,11 @@ func (s *SearchContext) CallDepth() string {
-	}
-
-	return false
+	return s.To.Contains(coverage)
```
👍

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -952,6 +952,21 @@ definitions:
       log:
         type: string
+  Port:
+    description: Layer 4 port/protocol tuple
```
Can we refer to "pair" instead? tuple is very networking heavy term and people draw conclusions. We don't want people to mistake this as both source destination tuple.

aanm replied:
Done

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -960,6 +975,16 @@ definitions:
       to:
         "$ref": "#/definitions/Labels"
+      l4-ingress:
+        description: List of l4 ingress rules
```
This needs a better description:
```
List of Layer 4 port and protocol pairs which the identity will use for outgoing network communication. If specified, all port/protocol pairs must be allowed by the policy.
```

aanm replied:
Just to be clear, you wrote `outgoing` should be `incoming`, right?

tgraf replied:
That's why ingress is not entirely clear :-) Is it ingress to the From or To?

tgraf commented on cilium/cmd/policy_trace.go:
```diff
@@ -25,15 +26,16 @@ import (
 // policyTraceCmd represents the policy_trace command
 var policyTraceCmd = &cobra.Command{
-	Use:   "trace -s <context> -d <context>",
+	Use:   "trace -s <context> -d <context> [--l4-in <port/protocol>] [--l4-out <port/protocol>]",
```
Let's reverse in and out so it is in the same order as `-s` and `-d`

aanm replied:
Fixed as discussed

tgraf commented on cilium/cmd/policy_trace.go:
```diff
@@ -46,9 +48,21 @@ SOURCE:KEY[=VALUE]`,
 
+		l4ingress, err := parseL4RulesSlice(l4ingress)
+		if err != nil {
+			Fatalf("Invalid ingress: %s", err)
```
"ingress" is very generic. Please include some L4 hint.

aanm replied:
Fixed

tgraf commented on cilium/cmd/policy_trace.go:
```diff
@@ -108,6 +124,33 @@ func parseAllowedSlice(slice []string) ([]string, error) {
+	for _, v := range slice {
+		vSplit := strings.Split(v, "/")
+		if len(vSplit) != 2 {
+			return nil, fmt.Errorf("invalid format %q. Should be <port>/<protocol>", v)
```
We should default to none as we do with policy. If not specified, both TCP and UDP will match.

aanm replied:
Done

tgraf commented on pkg/policy/rule_l4.go:
```diff
@@ -197,6 +216,25 @@ func (l4 L4PolicyMap) HasRedirect() bool {
+		return true
+	}
+	if len(l4Ports) == 0 {
+		return false
```
Why does this return false?

aanm replied:
I should probably change the name of the `containsAllL4` to a different one, suggestions needed.
The logic here is, we have an L4PolicyMap loaded, if the `l4Ports` is empty this means the L4 policy will be (and should be) denied for that search context. Makes sense?

tgraf commented on pkg/k8s/types/const.go:
```diff
@@ -0,0 +1,76 @@
+	DefaultPolicyParentPathPrefix = DefaultPolicyParentPath + NodePathDelimiter
+	// PodNamespaceLabel is the label used in kubernetes containers to
+	// specify which namespace they belong to.
+	PodNamespaceLabel      = "cilium.k8s.namespace"
```
Why is this not prefixed with `io.cilium`?

aanm replied:
Done. I was just trying to have small labels since the output can be `root.k8s.io.cilium.k8s.namespace`, it's easily changeable. ;-)

tgraf replied:
Why this long label? let's keep it simple. Does k8s have a standard label to indicate the namespace or are we adding that label?

aanm replied:
`root.k8s` is the "root" for the k8s policies, so far so good.
The `io.cilium.k8s.namespace` is our label to specify which namespace the pod is running. kubernetes has the `io.kubernetes.pod.namespace` which is longer, we can use the official one.
I was just trying to make similar labels for the namespace name (`cilium.k8s.namespace`) and the namespace labels (`cilium.k8s.ns-labels`)
```
$ cilium endpoint list
ENDPOINT   IDENTITY   LABELS (source:key[=value])                IPv6                     IPv4   STATUS
40447      410        k8s:k8s.cilium.k8s.namespace=kube-system   fd02::c0a8:210c:0:9dff          ready
                      k8s:k8s.cilium.k8s.ns-labels.admin=joe
                      k8s:k8s.k8s-app=kube-dns
```

tgraf replied:
The idea behind `root.k8s` is that it has the possibility to match on all labels due to the `coverAll` feature flag. So we can have rules which are attached to `root.k8s` which use the official `io.kubernetes.pod.namespace` label.

aanm replied:
We can, absolutely. I'll change it to `io.kubernetes.pod.namespace` ;-)

aanm committed:
policy: added Policy atomic replace operation.

This commit introduces a new operation type for the policy tree, it will
allow the daemon to modify the policy tree, by adding a subtracting
rules from 2 nodes in an atomic operation R/W operation.

AtomicReplace(subPath string, subNode *Node, addPath string, addNode *Node)
Examples:
tree
{
  "root": [{coverage: world, accept: id.foo},{coverage: host, accept: id.foo}],
  "children": ["id": [{coverage: foo, accept: root.bar}]
}
AtomicReplace("root", Node{coverage: world, accept: id.foo}, "", nil)
// Expected result:
tree
{
  "root": [{coverage: host, accept: id.foo}],
  "children": ["id": [{coverage: foo, accept: root.bar}]
}

AtomicReplace("root", Node{coverage: host, accept: id.foo},
              "root", Node{name: "id", coverage: foo, accept: id.foo})
// Expected result:
tree
{
  "root": [],
  "children": ["id": [{coverage: foo, accept: root.bar},
                      {coverage: foo, accept: id.foo}]
}

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
k8s: add namespace awareness to network policy

- Enforced namespace IngressIsolationPolicy
- Use all namespaces's labels in cilium endpoints from k8s.
- Changed valid prefixes to the new prefixes needed for by the
namespace IngressIsolationPolicy functionality.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
examples: update kubernetes example

Changed kubernetes example for the new functionalities introduced.

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on api/v1/models/port.go:
```diff
@@ -0,0 +1,83 @@
+)
+
+// prop value enum
+func (m *Port) validateProtocolEnum(path, location string, value string) error {
```
nit: ```... (path, location, value string ) ... ```

aanm replied:
@ianvernon `// This file was generated by the swagger tool.` :-)

aanm committed:
k8s: introduced cilium policy in k8s annotations

This commit will allow users to use cilium policy in k8s annotations.
This gives users the ability to allow incoming traffic to a particular
set of Pods from cilium reserved labels or even from endpoints running
outside the kubernetes cluster.

The cilium policy in the annotation will have the same structure as the
kubernetes network policy. The ingress.from.CiliumSelector from the
annotation will be merged into the spec.ingress.from.

Example of a policy with the cilium annotation:

---
apiVersion: extensions/v1beta1
kind: NetworkPolicy
metadata:
  name: guestbook-web
  annotations:
   cilium-policy: |
    {
      "ingress": [
        {
          "from": [
            {
              "ciliumSelector": {
                "matchLabels": {
                  "io.cilium.reserved": "world"
                }
              }
            }
          ]
        }
      ]
    }
spec:
  podSelector:
    matchLabels:
      k8s-app.guestbook: web
  ingress:
  - from:
    - podSelector: {}
    ports:
    - protocol: TCP
      port: 3000

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
k8s: refactoring k8s packages

This commit only moves some kubernetes constants and structures to a
dedicated k8s directory.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
Kubernetes network policy enhancements

- Introduced LabelSelector in Rule Consumers and L4 Rules. This allows
users to use the exact same semantics of K8s LabelSelector in Cilium.
- Changed the kubernetes network policy translation to make use of
the LabelSelector.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
policy: created function to check L4 policy

Moved the logic code from the policy handle to the tree package.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
start.sh: add missing worker's routes to master

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
policy: inherit IgnoreNameCoverage from children

If an orphan node is being added to the tree then its parents should
also contain the IgnoreNameCoverage by default.

Signed-off-by: André Martins <andre@cilium.io>
